[#67462] Add filterable tree view for parent selection

Kharonus opened the pull request:
# Ticket
[#67462](https://community.openproject.org/work_packages/67462)

# What are you trying to accomplish?
- change parent by selecting from existing hierarchy
  - show error if same item or same parent is selected, of if multiple items are selected

Kharonus committed:
[#67462] Add filterable tree view for parent selection

- https://community.openproject.org/work_packages/67462
- change parent by selecting from existing hierarchy
  - show error if same item or same parent is selected, of if multiple
    items are selected

NobodysNightmare commented on app/controllers/admin/custom_fields/hierarchy/items_controller.rb:
```diff
@@ -184,6 +199,24 @@ def add_errors_to_edit_form(validation_result)
+            new_parent = CustomField::Hierarchy::Item.including_children.find(input)
+            Success(new_parent)
+          else
+            Failure("Invalid input: #{new_parent_input}")
```
:red_circle:  Since you mentioned that case in the description, I tried selecting multiple new parents. I have to say that the error message surprised me:

<img width="1128" height="73" alt="image" src="https://github.com/user-attachments/assets/69700e62-be94-498e-8742-3adad21aaea9" />

Would it be possible for us to validate that case separately? Also: This error should be localized, shouldn't it?

Kharonus replied:
Yeah, I thought about it. Ideally, this one should be a programmatical error. Like the selector was wrongly configured. Sadly, for now, this is a common error.

My suggestion:
do not introduce a real error message, until we know, that it will be a user error in future, too. The feature is still behind a feature flag.

NobodysNightmare replied:
Mh. At first I wanted to disagree because it's a programmatical error in the frontend and the backend is never allowed to trust the frontend.

**But** we are properly catching the error here and the only problem is clarity and localization. I think that's fair for such a case.

NobodysNightmare commented on app/controllers/admin/custom_fields/hierarchy/items_controller.rb:
```diff
@@ -184,6 +199,24 @@ def add_errors_to_edit_form(validation_result)
+        end
+
+        def validate_new_parent(new_parent)
+          return Failure(I18n.t(:notice_change_parent_same_target)) if @active_item.parent.id == new_parent.id
```
:green_circle:  I am wondering whether this should be classified as an error condition. The state afterwards is exactly the state that the user wished for.

Mentally I compare this case to going into edit mode of the subject of a work package and then submitting the text that's already there. We also don't hint at the lack of change in that case.

Kharonus replied:
True, I tried to coalesce those handlings together, and when talking monads, you have either a success or a failure. This case is neither. I want to do nothing. Maybe I can amend the code a bit, to achieve that.

Kharonus replied:
As discussed before, I removed the error locale keys, as in a release ready version, those cases would be programmatical errors. The direct parent and the current item will be disabled for selection.

NobodysNightmare commented on app/components/admin/custom_fields/hierarchy/change_item_parent_dialog_component.html.erb:
```diff
@@ -28,18 +28,25 @@ See COPYRIGHT and LICENSE files for more details.
-          form.text_field(name: :new_parent, label: "New Parent")
+      primer_form_with(**form_arguments) do |form|
+        render(
+          Primer::OpenProject::FilterableTreeView.new(form_arguments: { builder: form, name: :new_parent })
```
The tree view includes some controls that seem absolutely useless for the current use case:

<img width="465" height="241" alt="image" src="https://github.com/user-attachments/assets/d9857d86-1c62-4ddf-8a95-b98b29b5cc6c" />

Is it possible to configure it to hide these controls? How hard would it be for us to extend it to hide these as per options?

(I am torn between rating this :yellow_circle:  or :green_circle: ... I find the useless controls extremely useless, though I see that we probably want to start working "with what we have")

Kharonus replied:
> Is it possible to configure it to hide these controls? How hard would it be for us to extend it to hide these as per options?

This is one of the feature requests for the primer team. We need a configuration flag to hide the segmented control - this is how it is called in the component.

NobodysNightmare commented on app/components/admin/custom_fields/hierarchy/change_item_parent_dialog_component.rb:
```diff
@@ -48,17 +48,48 @@ def form_id = "custom-fields--change-item-parent-form"
 
+        def hierarchy_items
+          hashed_hierarchy = @custom_field.hierarchy_root.hash_tree
+          hashed_hierarchy.nil? ? {} : hashed_hierarchy.first[1]
```
:question: That's a lot to unpack in a single line.

I understand that we return an empty hash if the hierarchy is nil (though I don't understand when it would be nil :one:).

But if it's present we take the first item. :two:  What is the first item?

And from that first item, we take the second item. :three:  What is that? :four: And which other items would it have?

I am wondering whether this would be more readable in the form of:

```ruby
return {} if hashed_hierarchy.nil?

wiggle = hashed_hierarchy.first
gadoozle = wiggle[1]
gadoozle
```

Though I guess that Rubocop would complain about a useless assignment in the assignment to `gadoozle`.

(In case it wasn't obvious: `wiggle` and `gadoozle` should be replaced with terms that explain what these things are)
```

Kharonus replied:
Okay, hint at readability granted :D 

This code is duplicated from the items component, were we construct a tree, too. I'd like to extract that code into a separate helper maybe? With project attributes we would even have a third place to unfold the hierarchy of a custom field and render a tree view from it - always with slight adjustments.

So, I will create a impl chunk to extract that code.

Kharonus replied:
See [chunk](https://community.openproject.org/wp/67544)

NobodysNightmare commented on app/components/admin/custom_fields/hierarchy/change_item_parent_dialog_component.rb:
```diff
@@ -48,17 +48,48 @@ def form_id = "custom-fields--change-item-parent-form"
+            else
+              expanded = current?(item) || child_hash.any? { |child, _| current?(child) }
+
+              tree.with_sub_tree(expanded: expanded, **item_options(item)) do |sub_tree|
```
Where is Rubocop?

:policeman: 

```suggestion
              tree.with_sub_tree(expanded:, **item_options(item)) do |sub_tree|
```

NobodysNightmare commented on app/controllers/admin/custom_fields/hierarchy/items_controller.rb:
```diff
@@ -102,12 +103,26 @@ def move
-          item_service.move_item(item: @active_item, new_parent:)
-
-          redirect_to(custom_field_item_path(@custom_field, new_parent), status: :see_other)
+        def change_parent # rubocop:disable Metrics/AbcSize
```
:yellow_circle: I hate it myself when Rubocop tells me about `Metrics/AbcSize` being to high on my method. But I am usually taking a look at whether useful extractions or simplifications to the method are possible.

I am not sure whether it will be enough to make Rubocop happy, but I can at least see a few chances that look useful to me, so I'd make those before declaring a method lost and disabling the cop:

* `params.require(...).require(...)` could be extracted into something like `new_parent_params`
* The assignment to `input` can then be avoided and we can directly call the new method in the line below
    * Alternatively we could maybe even call the new method directly from `parse_parent_input`... The method name tells me that it doesn't expect any other arguments anyways, so that might be valid
* I am still looking at that `either` call and I feel unhappy... but I can't quite make a suggestion how to fix it... Let me describe my feelings in another comment

NobodysNightmare commented on app/controllers/admin/custom_fields/hierarchy/items_controller.rb:
```diff
@@ -102,12 +103,26 @@ def move
+            validate_new_parent(new_parent).bind do
+              item_service.move_item(item: @active_item, new_parent:)
+            end
+          end.either(
```
:green_circle: I've been thinking why I find this `either` so hard to read. It might be a combination of things...

I am not the biggest fan of following `end` with `.` and here the thing that follows is a rather large method call... Maybe it would be easier to read if we did

```ruby
result = parse_parent_input(...) # the block from above goes here
result.either(
  ->(result)  do ... end, # now it's actually bad that this is called result too... maybe one of them could be renamed :see_no_evil: 
  ->(error) do ... end
)
```

(Note: This suggestion will _increase_ the Abc size, but I still wonder whether it increases readability)

Additionally I guess `either`'s readability is optimized for single line lambdas. The fact that we have two multi-line `do ... end` blocks inside a parameter list (i.e. `(...)`) that is attached to another `end` (see my point above) makes it so hard to unwrap for me.

(Marking this comment as green, because it's highly subjective)

Kharonus replied:
I still get that point. And usually I take the same approach, of trying to get rid of ABC size warning instead of disabling them. Taking another shot at this method.

NobodysNightmare commented on app/controllers/admin/custom_fields/hierarchy/items_controller.rb:
```diff
@@ -184,6 +199,24 @@ def add_errors_to_edit_form(validation_result)
+          case new_parent_input
+          in [new_parent]
+            input = MultiJson.load(new_parent, symbolize_keys: true)[:value]
+            new_parent = CustomField::Hierarchy::Item.including_children.find(input)
```
:yellow_circle: Should we explicitly handle the case of not finding the item with the given value?

Right now this case would raise a `RecordNotFound` error, when I would expect a `Failure` to be returned.

Current behaviour:

<img width="822" height="141" alt="image" src="https://github.com/user-attachments/assets/f07fa390-7b3f-41ab-a86e-860077869e97" />

Kharonus replied:
How can you trigger it? But in general, you are right. handling the error more gracefully can only help

NobodysNightmare replied:
Easy few steps :sunglasses: 

1. Open the assign new parent dialog and decide for a parent (no saving yet)
2. Open a new tab, delete the parent you decided for
3. Go back to the original tab and assign the parent

Kharonus replied:
true, very easy :D

NobodysNightmare reviewed (changes requested):
Mostly looking good. Though I think the error handling is not yet complete.

NobodysNightmare commented:
Question from the demo today: Is it possible to move an item to the root of the tree?

Kharonus commented:
> Question from the demo today: Is it possible to move an item to the root of the tree?

not yet, but this must be possible. Parimal is looking into it.

Kharonus committed:
Merge branch 'dev' into implementation/67462-add-filterable-tree-view-component-to-dialog

Kharonus committed:
[#67462] code cleanup after code review

- removed locale keys that would never be caused by user error
- display root element in change parent dialog

NobodysNightmare commented on app/controllers/admin/custom_fields/hierarchy/items_controller.rb:
```diff
@@ -184,6 +203,29 @@ def add_errors_to_edit_form(validation_result)
+            if new_parent.present?
+              Success(new_parent)
+            else
+              Failure("Cannot find parent with id: #{input}")
```
:green_circle: I'd still argue that this is an error that can happen without a programming error. As shown in the reproduction steps in my earlier comment, all you need is a human deleting the parent that you want to move to.

--> I'd probably translate here

(though the likelyhood to see this error in practice is pretty low)

Kharonus replied:
yes, true, for this one I agree. Will fix this in the next PR.
